grpc/stream.go

```diff
@@ -0,0 +1,344 @@
+	for {
+		msg, err := s.stream.ReceiveConverted()
+		if errors.Is(err, grpcext.ErrCanceled) {
+			s.logger.Debug("stream is cancelled")
+
+			err = nil
+		}
+
+		if err != nil && !errors.Is(err, io.EOF) {
+			s.logger.WithError(err).Debug("error while reading from the stream")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(err)
+			})
+
+			return
+		}
+
+		// no message means that the stream has been closed
+		if len(msg) == 0 && errors.Is(err, io.EOF) {
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+
+			return
+		}
+
+		select {
+		case readDataChan <- msg:
+		case <-s.done:
+			if len(msg) > 0 {
+				s.logger.Debug("stream is done, sending the last message")
+
+				readDataChan <- msg
+			}
+
+			return
+		}
+	}
+}
+
+// writeData writes data to the stream
+func (s *stream) writeData() {
+	writeChannel := make(chan message)
+	go func() {
+		for {
+			select {
+			case msg, ok := <-writeChannel:
+				if !ok {
+					return
+				}
+
+				err := s.stream.Send(msg.msg)
+				if err != nil {
+					s.logger.WithError(err).Error("failed to send data to the stream")
+
+					s.tq.Queue(func() error {
+						closeErr := s.closeWithError(err)
+						return closeErr
+					})
+					return
+				}
+
+			case <-s.done:
+				return
+			}
+		}
+	}()
+
+	{
+		defer func() {
+			close(writeChannel)
+		}()
+
+		queue := make([]message, 0)
+		var wch chan message
+		var msg message
+		for {
+			wch = nil // this way if nothing to read it will just block
+			if len(queue) > 0 {
+				msg = queue[0]
+				wch = writeChannel
+			}
+			select {
+			case msg = <-s.writeQueueCh:
+				select {
+				case writeChannel <- msg:
+				default:
+					queue = append(queue, msg)
+				}
+			case wch <- msg:
+				queue = queue[:copy(queue, queue[1:])]
+
+			case <-s.done:
+				return
+			}
+		}
+	}
+}
+
+// on registers a listener for a certain event type
+func (s *stream) on(event string, listener func(goja.Value) (goja.Value, error)) {
+	if err := s.eventListeners.add(event, listener); err != nil {
+		s.vu.State().Logger.Warnf("can't register %s event handler: %s", event, err)
+	}
+}
+
+// write writes a message to the stream
+func (s *stream) write(input goja.Value) {
+	if input == nil || goja.IsNull(input) || goja.IsUndefined(input) {
+		s.logger.Warnf("can't send empty message")
+	}
+
+	rt := s.vu.Runtime()
+
+	b, err := input.ToObject(rt).MarshalJSON()
+	if err != nil {
+		s.logger.WithError(err).Warnf("can't marshal message")
+	}
+
+	s.writeQueueCh <- message{msg: b}
+}
+
+// end closes client the stream
+func (s *stream) end() {
+	var err error
+
+	defer func() {
+		_ = s.closeWithError(err)
+	}()
+
+	if err = s.stream.CloseSend(); err != nil {
+		s.logger.WithError(err).Error("failed to close the stream")
+
+		return
+	}
+
+	// TODO: move this ?
+	_ = s.callEventListeners(eventEnd)
+}
+
+func (s *stream) closeWithError(err error) error {
+	if s.closed {
+		s.logger.WithError(err).Debug("connection is already closed")
+
+		return nil
+	}
```

Comment:
Do we need one more boolean? We could check the chan directly

Reply:
:thinking: 

>We could check the chan directly

By checking the channel directly, do you mean an attempt to read it? :thinking:

Reply:
```suggestion
	select {
	    case <-done:
		s.logger.WithError(err).Debug("connection is already closed")
		return nil
            default:
	}
```
If I'm getting right we are not invoking this function with a high rate, so it should be fine having the select overhead but it keeps the logic consistent and keep the source of truth on the channel.